ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRelFieldTrimmer.java

```diff
@@ -502,28 +502,15 @@ private ImmutableBitSet generateNewGroupset(Aggregate aggregate, ImmutableBitSet
       return generateGroupSetIfCardinalitySame(aggregate, originalGroupSet, fieldsUsed);
     }
 
-    // we have set of unique key, get to the key which is same as group by key
-    ImmutableBitSet groupByUniqueKey = null;
-
+    // Find the maximum number of columns that can be removed by retaining a certain unique key
+    ImmutableBitSet columnsToRemove = ImmutableBitSet.of();
     for (ImmutableBitSet key : uniqueKeys) {
-      if (aggregate.getGroupSet().contains(key)) {
-        groupByUniqueKey = key;
-        break;
+      ImmutableBitSet removableCols = originalGroupSet.except(key).except(fieldsUsed);
```

Comment:
Is it possible to compute `except(fieldsUsed)` outside the loop? I believe changing the order of `except`s will still yield the same resulting set.

Reply:
Yes we can do this, but just want to be careful that if there is no uniquekey match then we should not remove any columns at all(should skip the except in the return statement below in that case)

Reply:
Just realized that we only update columnsToRemove when there is atleast one match. So I think yes we can move the except(fieldsUsed) outside the for loop.
Also is it efficient to loop the fieldsUsed(and then check if there is a field that is a unique key and part of aggregate keys to retain only fieldsUsed) vs looping uniquekeys? Mostly depends on size of fieldsUsed vs uniqueKeys. I will probably leave it upto you to decide on that.

Reply:
Yes, it is possible to compute `except(fieldsUsed)` outside the loop. I applied the suggestion in https://github.com/apache/hive/pull/5586/commits/e89386dc64bd4eb8a3981795b3cacd264861ee35.

Reply:
Comparing the efficiency of the iteration between `fieldsUsed` and `uniqueKeys` is not possible because the semantics are different. The `uniqueKeys` variable is a set of sets (`Set<ImmutableBitSet>`) while `fieldsUsed` is a single set (`ImmutableBitSet`). Note that a `key` is not necessarily a single column but a set of columns (composite key).

Reply:
Thank you for your explanation. Makes sense to me.